tests/layers/tuple_matchers/threshold_tuple_matcher_test.py

```diff
@@ -0,0 +1,140 @@
+                                  name="input_tuple2")
+        self.num_hidden_layers = 1
+        self.hidden_layer_width = 2
+        self.hidden_layer_activation = 'linear'
+
+        # tuple1 shape: (batch size, num_slots, num_words, embed_dimension)
+        self.tuple1 = numpy.random.rand(1, num_slots, num_words, embed_dimension)
+        # tuple2 shape: (batch size, num_slots, num_words, embed_dimension)
+        self.tuple2 = numpy.random.rand(1, num_slots, num_words, embed_dimension)
+        # cause some dimensions to have identical embeddings (i.e. cosine similarity of 1)
+        self.tuple1[0, 1, 1, :] = numpy.ones(4)
+        self.tuple1[0, 2, 0, :] = numpy.ones(4)
+
+        self.tuple2[0, 1, 0, :] = numpy.ones(4)
+        self.tuple2[0, 2, 3, :] = numpy.ones(4)
+        self.tuple2[0, 2, 4, :] = numpy.ones(4)
+        # This should result in 1 matched word in slot 1, and two in slot 2
+        # So the normalized overlaps should be (0, 1/5,  2/5)
+
+    def test_general_case(self):
+        # Custom initialization for this test to account for rounding in the cosine similarity.
+        # pylint: disable=unused-argument
+        def custom_initialization(shape, name=None):
+            return K.ones(shape) * 0.999
+
+        match_layer = ThresholdTupleMatcher({"type": "cosine_similarity"},
+                                            self.num_hidden_layers,
+                                            self.hidden_layer_width,
+                                            initialization=custom_initialization,
+                                            hidden_layer_activation=self.hidden_layer_activation)
+        output = match_layer([self.tuple1_input, self.tuple2_input])
+        model = Model([self.tuple1_input, self.tuple2_input], output)
+
+        # Get the initial weights for use in testing
+        layer_nn = match_layer.hidden_layer_weights
+
+        # Testing general unmasked case.
+        desired_overlap = K.variable(numpy.asarray([[0, 1/5, 2/5]]))
+        # Desired_overlap gets fed into the inner NN.
+        neural_network_feed_forward = apply_feed_forward(desired_overlap, layer_nn,
+                                                         activations.get(match_layer.hidden_layer_activation))
+        # Apply the final activation function.
+        desired_result = activations.get(match_layer.final_activation)(K.dot(neural_network_feed_forward,
+                                                                             match_layer.score_layer))
+        result = model.predict([self.tuple1, self.tuple2])
+        assert_array_almost_equal(result, K.eval(desired_result))
+
+
+    def test_returns_masks_correctly(self):
+        # Test when one tuple is all padding.
+        # Here, since tuple3 is all padding, we want to return a mask value of 0 for this pair
+        tuple1 = K.variable(self.tuple1)
+        mask1 = K.variable(numpy.asarray([[1, 1, 1]]))
```

Comment:
right you are.  thanks for the in-depth code review, matt! fixed and as you predicted, test failed, fixed the compute_mask as you did in word_overlap tuple matcher (NOW i better understand *why* you did that fix too), and we're now passing, but please do check again when I re-submit this sucker!